File: extractor.py
import httpx
from google import genai
from google.genai import types
from google.genai import errors as genai_errors
import json
import logging
import os
from dotenv import load_dotenv
from pydantic import ValidationError
from schemas import InvoiceExtraction

load_dotenv()

logger = logging.getLogger(__name__)

client = genai.Client(api_key=os.getenv("GEMINI_API_KEY"))
# creates the Gemini client — like new Anthropic(api_key=...) in lesson 2.2

MAX_FILE_SIZE_BYTES = 20 * 1024 * 1024  # 20 MB
PDF_MAGIC_BYTES = b"%PDF-"


class ExtractionError(Exception):
    # sanitized, user-facing error — callers (e.g. the API route) can
    # show str(e) to clients without leaking internal details
    pass


def extract_invoice(file_bytes: bytes, mime_type: str) -> dict:
    # file_bytes — raw binary content of the PDF file
    # mime_type — "application/pdf"
    # returns a plain dict with the extracted invoice data

    if mime_type != "application/pdf":
        raise ExtractionError("Only application/pdf is supported")

    if not file_bytes:
        raise ExtractionError("Empty file")

    if len(file_bytes) > MAX_FILE_SIZE_BYTES:
        raise ExtractionError(
            f"File too large: max {MAX_FILE_SIZE_BYTES // (1024 * 1024)}MB"
        )

    if not file_bytes.startswith(PDF_MAGIC_BYTES):
        raise ExtractionError("File is not a valid PDF")

    schema_json = json.dumps(
        InvoiceExtraction.model_json_schema(),
        ensure_ascii=False
        # ensure_ascii=False — keeps Hebrew text readable (lesson 2.4)
    )

    prompt = f"""You are an invoice data extraction system.
    Extract all data from this invoice document.

    Return a JSON object matching this schema exactly:
    {schema_json}

    Rules:
    - All monetary values must be in ILS (convert if needed)
    - vat_rate should be 0.17 for Israeli 17% VAT
    - If a field is missing from the invoice, use null
    - confidence: your certainty that all extracted data is correct (0.0 to 1.0)
    - Return ONLY valid JSON, no explanation text"""

    file_part = types.Part.from_bytes(data=file_bytes, mime_type=mime_type)
    # wraps the file bytes correctly for the new google-genai SDK

    try:
        response = client.models.generate_content(
            model="gemini-2.5-flash",
            contents=[file_part, prompt],
            # sends BOTH the file and the prompt together
            config={"response_mime_type": "application/json"}
            # forces JSON output — no extra text around it
        )
    except (genai_errors.APIError, httpx.HTTPError) as e:
        logger.exception("Gemini API call failed")
        raise ExtractionError("Extraction service is unavailable") from e

    response_text = response.text
    if not response_text:
        logger.error("Gemini returned an empty response")
        raise ExtractionError("Extraction produced no data")

    try:
        raw_data = json.loads(response_text)
        # json.loads() — parses JSON string → dict, like JSON.parse() in JS
    except json.JSONDecodeError as e:
        logger.exception("Gemini response was not valid JSON: %r", response_text)
        raise ExtractionError("Extraction returned malformed data") from e

    try:
        validated = InvoiceExtraction(**raw_data)
        # creates Pydantic instance and validates all fields
        # ** unpacks dict as arguments: func(**{"a": 1}) = func(a=1)
        # throws an error if any field is missing or wrong type
    except ValidationError as e:
        logger.exception("Gemini response failed schema validation: %r", raw_data)
        raise ExtractionError("Extraction result did not match expected format") from e

    return validated.model_dump(mode="json")
    # converts Pydantic instance back to plain dict
    # mode="json" — converts date objects to strings like "2026-06-01"

File: web/lib/extractor.ts
import { GoogleGenAI } from "@google/genai";

const MAX_FILE_SIZE_BYTES = 20 * 1024 * 1024; // 20 MB
const PDF_MAGIC_BYTES = Buffer.from("%PDF-");
const MODEL = "gemini-2.5-flash";
const GEMINI_TIMEOUT_MS = 15_000;

export class ExtractionError extends Error {}

export interface InvoiceItem {
  description: string;
  quantity: number;
  unit_price: number;
  line_total: number;
}

export interface InvoiceExtraction {
  vendor: string;
  vendor_id: string | null;
  invoice_number: string | null;
  invoice_date: string | null;
  due_date: string | null;
  items: InvoiceItem[];
  subtotal: number;
  vat_rate: number;
  vat_amount: number;
  total: number;
  currency: string;
  confidence: number;
}

const client = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });

export function isValidInvoiceItem(value: unknown): value is InvoiceItem {
  if (typeof value !== "object" || value === null) return false;
  const item = value as Record<string, unknown>;
  return (
    typeof item.description === "string" &&
    typeof item.quantity === "number" &&
    typeof item.unit_price === "number" &&
    typeof item.line_total === "number"
  );
}

export function validateInvoiceExtraction(data: unknown): InvoiceExtraction {
  if (typeof data !== "object" || data === null) {
    throw new ExtractionError("Extraction result was not a JSON object");
  }
  const d = data as Record<string, unknown>;

  const isNullableString = (v: unknown) => v === null || typeof v === "string";

  if (typeof d.vendor !== "string" || d.vendor.trim() === "") {
    throw new ExtractionError("Extraction result is missing 'vendor'");
  }
  if (!isNullableString(d.vendor_id)) {
    throw new ExtractionError("Extraction result has an invalid 'vendor_id'");
  }
  if (!isNullableString(d.invoice_number)) {
    throw new ExtractionError("Extraction result has an invalid 'invoice_number'");
  }
  if (!isNullableString(d.invoice_date)) {
    throw new ExtractionError("Extraction result has an invalid 'invoice_date'");
  }
  if (!isNullableString(d.due_date)) {
    throw new ExtractionError("Extraction result has an invalid 'due_date'");
  }
  if (!Array.isArray(d.items) || !d.items.every(isValidInvoiceItem)) {
    throw new ExtractionError("Extraction result has invalid or missing 'items'");
  }
  for (const field of ["subtotal", "vat_rate", "vat_amount", "total", "confidence"] as const) {
    if (typeof d[field] !== "number") {
      throw new ExtractionError(`Extraction result has an invalid '${field}'`);
    }
  }
  if (typeof d.currency !== "string") {
    throw new ExtractionError("Extraction result has an invalid 'currency'");
  }

  return d as unknown as InvoiceExtraction;
}

export const PROMPT = `You are an invoice data extraction system.
Extract all data from this invoice document.

Return a JSON object with exactly these fields:
- vendor (string): company or person who issued the invoice. If the name
  appears in both Hebrew and English, extract the Hebrew name — always use
  the same language for the same company so results are consistent.
- vendor_id (string or null): Israeli business ID (ח.פ. or ע.מ.), null if not found
- invoice_number (string or null): invoice number or reference ID
- invoice_date (string or null): date the invoice was issued, format YYYY-MM-DD
- due_date (string or null): payment due date, null if not specified, format YYYY-MM-DD
- items (array): list of line items, each with:
  - description (string): product or service name
  - quantity (number): number of units
  - unit_price (number): price per single unit in ILS, before VAT
  - line_total (number): quantity * unit_price for this line item
- subtotal (number): total before VAT in ILS
- vat_rate (number): VAT percentage as decimal, e.g. 0.17 for 17%
- vat_amount (number): VAT amount in ILS
- total (number): final total including VAT in ILS
- currency (string): currency code, e.g. ILS, USD, EUR
- confidence (number): 0.0 to 1.0, your certainty that all extracted data is correct

Rules:
- All monetary values must be in ILS (convert if needed)
- vat_rate should be 0.17 for Israeli 17% VAT
- If a field is missing from the invoice, use null
- Return ONLY valid JSON, no explanation text`;

export async function extractInvoice(
  fileBytes: Buffer,
  mimeType: string
): Promise<InvoiceExtraction> {
  if (mimeType !== "application/pdf") {
    throw new ExtractionError("Only application/pdf is supported");
  }

  if (!fileBytes || fileBytes.length === 0) {
    throw new ExtractionError("Empty file");
  }

  if (fileBytes.length > MAX_FILE_SIZE_BYTES) {
    throw new ExtractionError(
      `File too large: max ${MAX_FILE_SIZE_BYTES / (1024 * 1024)}MB`
    );
  }

  if (!fileBytes.subarray(0, PDF_MAGIC_BYTES.length).equals(PDF_MAGIC_BYTES)) {
    throw new ExtractionError("File is not a valid PDF");
  }

  let responseText: string | undefined;
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), GEMINI_TIMEOUT_MS);
  try {
    try {
      const response = await client.models.generateContent({
        model: MODEL,
        contents: [
          { inlineData: { data: fileBytes.toString("base64"), mimeType } },
          { text: PROMPT },
        ],
        config: {
          responseMimeType: "application/json",
          abortSignal: controller.signal,
        },
      });
      responseText = response.text;
    } finally {
      clearTimeout(timeout);
    }
  } catch (err) {
    console.error("Gemini extraction call failed:", err);
    throw new ExtractionError(
      err instanceof Error && err.name === "AbortError"
        ? `Extraction timed out after ${GEMINI_TIMEOUT_MS / 1000}s`
        : "Extraction service is unavailable"
    );
  }

  if (!responseText) {
    throw new ExtractionError("Extraction produced no data");
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(responseText);
  } catch (err) {
    console.error("Gemini response was not valid JSON:", err, responseText);
    throw new ExtractionError("Extraction returned malformed data");
  }

  return validateInvoiceExtraction(parsed);
}
